Allow constructing MicroHTTP with any implementor of ToSocketAddrs and replace usage of try! macro with ?

chili-b opened the pull request:
Modify the signature of the new() function of MicroHTTP to be a generic
accepting any implementor of std::net::ToSocketAddrs instead of &str.
Also, all usage of the outdated try! macro has been replace with ? to
resolve warnings on compilation.

&str implements the ToSocketAddrs trait, so all test still pass with
zero modification and this change will not break compatibility with
usage assuming the previous API.

This is a great little library and exactly what I was looking for in terms
of scope. Thank you for making it!

dexgs committed:
Allow constructing MicroHTTP with any implementor of ToSocketAddrs and
replace usage of try! macro with ?

Modify the signature of the new() function of MicroHTTP to be a generic
accepting any implementor of std::net::ToSocketAddrs instead of &str.
Also, all usage of the outdated try! macro has been replace with ? to
resolve warnings on compilation.

dexgs committed:
Add methods to respond with types implementing Read

Previously, it was only possible to respond with a
byte slice (&[u8]) which is suboptimal for responding
to a request with the contents of a somewhat large file
as it would require loading the whole thing into memory.

The added methods allow a file-backed response's body
to be read in 4096 byte chunks instead of all at once.

ghost commented:
I just added a commit which adds methods to respond with a type implementing the Read trait which is useful for sending files to a client without having to load the whole thing into memory. I again left the original API unchanged so as not to break compatibility with anything using the current API. I documented the new methods, but they may not be up to your standards. I will change anything if needed.

dexgs committed:
Move setting nonblocking on MicroHTTP to its own method

This allows a choice instead of forcing nonblocking which
results in high CPU usage if you wait for connections in a loop.

philippludwig commented:
Thank you. For some reason I did not get a notification for this PR.

philippludwig commented:
And let me add that I appreciate your contribution; I am sorry for not seeing this earlier.
